src/jit/emit.cpp

```diff
@@ -3793,6 +3801,42 @@ void                emitter::emitJumpDistBind()
 
         // If this is a jump via register, the instruction size does not change, so we are done.
 
+#if defined(_TARGET_ARM64_)
+        // JIT code and data will be allocated together for arm64 so the relative offset to JIT data is known.
+        // In case such offset can be encodeable for `ldr` (+-1MB), shorten it.
+        if (emitIsLoadConstant(jmp))
+        {
+            // Reference to JIT data
+            assert(jmp->idAddr()->iiaIsJitDataOffset());
+            assert(jmp->idIsBound());
+            UNATIVE_OFFSET srcOffs = jmpIG->igOffs + jmp->idjOffs;
+
+            int doff = jmp->idAddr()->iiaGetJitDataOffset();
+            assert(doff >= 0);
+            ssize_t imm = emitGetInsSC(jmp);
+            assert((imm >= 0) && (imm < 0x1000));   // 0x1000 is arbitrary, currently 'imm' is always 0
+
+            unsigned dataOffs = (unsigned)(doff + imm);
+            assert(dataOffs < emitDataSize());
+
+            // Conservately assume JIT data starts after the entire code size.
+            // TODO: we might consider only hot code size which will be computed later in emitComputeCodeSizes().
```

Comment:
We are already allocating read-only data next to the hot code, and in this change reference from cold code is always kept long. The only minor issue is that emitComputeCodeSizes() comes later than this although I can replicate the code to get hot code size. I'd like to reorder emitComputeCodeSizes() with emitJumpDistBind(), which appears okay, but given the risk of changing the common code toward RTM, I will put ARM64-TODO and follow it up later.